File: Codeforces Round #837 (Div. 2) /B.cpp
// https://codeforces.com/contest/1771/problem/B

#include <assert.h>
#include <bits/stdc++.h>

using namespace std;

typedef unsigned int uint;

struct edge {
  uint x, y;
};

static inline tuple<uint, vector<struct edge>> parse1(void) {
  uint n, m;
  cin >> n >> m;

  vector<struct edge> edges(m);
  for (uint i = 0; i < m; ++i) {
    uint x, y;
    cin >> x >> y;
    edges[i] = {x, y};
  }

  return {n, edges};
}

static inline uint solve1(uint n, vector<struct edge> edges) {

  map<uint, set<uint>> adj;
  for (struct edge e : edges) {
    adj[e.x].insert(e.y);
    adj[e.y].insert(e.x);
  }

  uint res = 0;
  uint racha_starts_at = 1;
  for (uint i = 1; i <= n; i++) {
    set<uint> adj_i = adj[i];

    set<uint>::iterator start = adj_i.lower_bound(racha_starts_at);
    set<uint>::iterator end = adj_i.upper_bound(i-1);

    bool breaks_racha = start != end && start != adj_i.end();

    if (breaks_racha) {
      racha_starts_at = *prev(end) + 1;
    }

    res += i - racha_starts_at + 1;
  }

  return res;
}

int main(void) {
  ios_base::sync_with_stdio(false);
  cin.tie(NULL);

  uint t;
  cin >> t;
  for (uint i = 0; i < t; ++i) {
    uint n;
    vector<struct edge> edges;
    tie(n, edges) = parse1();
    cout << solve1(n, edges) << "\n";
  }

  return EXIT_SUCCESS;
}

File: Codeforces Round #837 (Div. 2) /B.py
# https://codeforces.com/contest/1771/problem/B

from typing import Dict, List, Tuple


def solve1(n: int, As: List[Tuple[int, int]]) -> int:
    d: Dict[int, List[int]] = {}

    for (x, y) in As:
        if x not in d:
            d[x] = []
        d[x].append(y)

        if y not in d:
            d[y] = []
        d[y].append(x)

    res: int = 0
    i: int = 1
    racha_starts_at: int = 1
    while i <= n:
        # print(f"{i=}")
        d_i: List[int] = []
        if i in d:
            d_i = d[i]

        breaks_racha: bool = False
        j: int = i-1
        while j >= racha_starts_at:
            if j in d_i:
                breaks_racha = True
                break
            j -= 1

        if breaks_racha:
            #print("breaks racha")
            racha_starts_at = j + 1
            # print(f"{racha_starts_at=}")
            res -= int((i - racha_starts_at) * (i - racha_starts_at+1)/2)
            i = racha_starts_at
        else:
            res += i - racha_starts_at + 1
            i += 1

        # print(f"{res=}")

    return res


def parse1() -> Tuple[int, List[Tuple[int, int]]]:
    n, m = (int(x) for x in input().split())
    As = [tuple(int(x) for x in input().split()) for _ in range(m)]
    return n, As


def main():
    t = int(input())
    for _ in range(t):
        print(solve1(*parse1()))


if __name__ == '__main__':
    main()
